### PR TITLE
ict: improve log configuration - add options '--verbose' and '--trace'

### PR DESCRIPTION
**Pull request by jwausle**
ict: improve log configuration - add options '--verbose' and '--trace'

* enhance cmdline to configure log.level via '--verbose' and '--trace'
* enable rolling log file 'logs/ict.log' when cmdline option '--logfile-enabled'
* remove unused 'utils.ErrorHandler'

### CODE REVIEW

File: src/test/java/org/iota/ict/MainLog4jConfigTest.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            " logDir=logs/,\n" +
+            " logFilename=ict.log,\n" +
+            " logFileEnabled=false\n" +
+            "}";
+    private static final List<String> KNOWING_LOG4J_CONTEXT_VARIABLES = Arrays.asList("logLevelSplit", "logFileEnabled", "logDir", "logFilename");
+
+    private final Main.Log4JConfig underTest = Main.Log4JConfig.getDefault();
+
+    @Test
+    public void default_toString_test() {
```
[the remark this answers is not in the document]
Comment (jwausle) [reply src/test/java/org/iota/ict/MainLog4jConfigTest.java]: Fixed.

File: src/main/java/org/iota/ict/Main.java [hunk 2]
```diff
@@ -212,14 +236,90 @@ private String parseValueOrNull(String[] args, String... argsFilter) {
...
+        private final boolean disableRootLevelSetting;
+        private Level rootLevel = Level.INFO;
+
+        static Log4JConfig getDefault() {
+            return new Log4JConfig();
+        }
+
+        private Log4JConfig() {
+            disableRootLevelSetting = System.getProperties().containsKey("log4j.configurationFile");
+            ThreadContext.put(CTX_KEY_LOG_LEVEL_SPLIT, Level.WARN.name());
```
[the remark this answers is not in the document]
Comment (jwausle) [reply src/main/java/org/iota/ict/Main.java]: Removed.

File: src/test/java/org/iota/ict/MainLog4jConfigTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikrohash) [reply src/test/java/org/iota/ict/MainLog4jConfigTest.java]: Thanks for the suggestion. That would definitely make sense.

**Comment by JelleMillenaar** [comment]
ICT crashes on startup using Windows 10 due to the log4j filename not being correct. 

Initial error:
2019-03-09 15:23:24,205 org.iota.ict.Ict ERROR Unable to create file ${ctx:logDir}/${ctx:logFilename} java.io.IOException: The filename, directory name, or volume label syntax is incorrect

**Comment by jwausle** [comment]
@JelleMillenaar Could you provide ict-version + the full cmdline + full stacktrace for reproduction.

**Comment by mikrohash** [comment]
When starting Ict on Windows 10, I get a lot of these:

```
2019-03-22 11:46:20,853 main ERROR Unable to create file ${ctx:logDir}/${ctx:logFilename} java.io.IOException: Die Syntax für den Dateinamen, Verzeichnisnamen oder die Datenträgerbezeichnung ist falsch
	at java.io.WinNTFileSystem.canonicalize0(Native Method)
	at java.io.WinNTFileSystem.canonicalize(WinNTFileSystem.java:428)
	at java.io.File.getCanonicalPath(File.java:618)
	at java.io.File.getCanonicalFile(File.java:643)
	at org.apache.logging.log4j.core.util.FileUtils.makeParentDirs(FileUtils.java:142)
	at org.apache.logging.log4j.core.appender.rolling.RollingFileManager$RollingFileManagerFactory.createManager(RollingFileManager.java:627)
	at org.apache.logging.log4j.core.appender.rolling.RollingFileManager$RollingFileManagerFactory.createManager(RollingFileManager.java:608)
	at org.apache.logging.log4j.core.appender.AbstractManager.getManager(AbstractManager.java:113)
	at org.apache.logging.log4j.core.appender.OutputStreamManager.getManager(OutputStreamManager.java:114)
	at org.apache.logging.log4j.core.appender.rolling.RollingFileManager.getFileManager(RollingFileManager.java:188)
	at org.apache.logging.log4j.core.appender.RollingFileAppender$Builder.build(RollingFileAppender.java:145)
	at org.apache.logging.log4j.core.appender.RollingFileAppender$Builder.build(RollingFileAppender.java:61)
	at org.apache.logging.log4j.core.config.plugins.util.PluginBuilder.build(PluginBuilder.java:123)
	at org.apache.logging.log4j.core.config.AbstractConfiguration.createPluginObject(AbstractConfiguration.java:959)
	at org.apache.logging.log4j.core.config.AbstractConfiguration.createConfiguration(AbstractConfiguration.java:899)
	at org.apache.logging.log4j.core.appender.routing.RoutingAppender.createAppender(RoutingAppender.java:271)
	at org.apache.logging.log4j.core.appender.routing.RoutingAppender.getControl(RoutingAppender.java:255)
	at org.apache.logging.log4j.core.appender.routing.RoutingAppender.append(RoutingAppender.java:225)
	at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:156)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:129)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:120)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:84)
	at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:464)
	at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:448)
	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:431)
	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:406)
	at org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(AwaitCompletionReliabilityStrategy.java:63)
	at org.apache.logging.log4j.core.Logger.logMessage(Logger.java:146)
	at org.apache.logging.log4j.spi.AbstractLogger.tryLogMessage(AbstractLogger.java:2170)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessageTrackRecursion(AbstractLogger.java:2125)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessageSafely(AbstractLogger.java:2108)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:2002)
	at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:1974)
	at org.apache.logging.log4j.spi.AbstractLogger.info(AbstractLogger.java:1311)
	at org.iota.ict.api.RestApi.onStart(RestApi.java:119)
	at org.iota.ict.utils.RestartableThread$StateTerminated.start(RestartableThread.java:76)
	at org.iota.ict.utils.RestartableThread.start(RestartableThread.java:37)
	at org.iota.ict.utils.RestartableThread$StateTerminated.start(RestartableThread.java:78)
	at org.iota.ict.utils.RestartableThread.start(RestartableThread.java:37)
	at org.iota.ict.Ict.<init>(Ict.java:70)

2019-03-22 11:46:20,871 main ERROR Could not create plugin of type class org.apache.logging.log4j.core.appender.RollingFileAppender for element RollingFile: java.lang.IllegalStateException: ManagerFactory [org.apache.logging.log4j.core.appender.rolling.RollingFileManager$RollingFileManagerFactory@10db82ae] unable to create manager for [${ctx:logDir}/${ctx:logFilename}] with data [org.apache.logging.log4j.core.appender.rolling.RollingFileManager$FactoryData@501edcf1[pattern=${ctx:logDir}/${ctx:logFilename}-%d{MM-dd-yyyy}.gz, append=true, bufferedIO=true, bufferSize=8192, policy=TimeBasedTriggeringPolicy(nextRolloverMillis=0, interval=1, modulate=false), strategy=DefaultRolloverStrategy(min=1, max=7, useMax=true), advertiseURI=null, layout=%d{HH:mm:ss.SSS} %-5level [%t/%logger{0}]   %msg%n%throwable, filePermissions=null, fileOwner=null]] java.lang.IllegalStateException: ManagerFactory [org.apache.logging.log4j.core.appender.rolling.RollingFileManager$RollingFileManagerFactory@10db82ae] unable to create manager for [${ctx:logDir}/${ctx:logFilename}] with data [org.apache.logging.log4j.core.appender.rolling.RollingFileManager$FactoryData@501edcf1[pattern=${ctx:logDir}/${ctx:logFilename}-%d{MM-dd-yyyy}.gz, append=true, bufferedIO=true, bufferSize=8192, policy=TimeBasedTriggeringPolicy(nextRolloverMillis=0, interval=1, modulate=false), strategy=DefaultRolloverStrategy(min=1, max=7, useMax=true), advertiseURI=null, layout=%d{HH:mm:ss.SSS} %-5level [%t/%logger{0}]   %msg%n%throwable, filePermissions=null, fileOwner=null]]
	at org.apache.logging.log4j.core.appender.AbstractManager.getManager(AbstractManager.java:115)
	at org.apache.logging.log4j.core.appender.OutputStreamManager.getManager(OutputStreamManager.java:114)
	at org.apache.logging.log4j.core.appender.rolling.RollingFileManager.getFileManager(RollingFileManager.java:188)
	at org.apache.logging.log4j.core.appender.RollingFileAppender$Builder.build(RollingFileAppender.java:145)
	at org.apache.logging.log4j.core.appender.RollingFileAppender$Builder.build(RollingFileAppender.java:61)
	at org.apache.logging.log4j.core.config.plugins.util.PluginBuilder.build(PluginBuilder.java:123)
	at org.apache.logging.log4j.core.config.AbstractConfiguration.createPluginObject(AbstractConfiguration.java:959)
	at org.apache.logging.log4j.core.config.AbstractConfiguration.createConfiguration(AbstractConfiguration.java:899)
	at org.apache.logging.log4j.core.appender.routing.RoutingAppender.createAppender(RoutingAppender.java:271)
	at org.apache.logging.log4j.core.appender.routing.RoutingAppender.getControl(RoutingAppender.java:255)
	at org.apache.logging.log4j.core.appender.routing.RoutingAppender.append(RoutingAppender.java:225)
	at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:156)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:129)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:120)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:84)
	at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:464)
	at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:448)
	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:431)
	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:406)
	at org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(AwaitCompletionReliabilityStrategy.java:63)
	at org.apache.logging.log4j.core.Logger.logMessage(Logger.java:146)
	at org.apache.logging.log4j.spi.AbstractLogger.tryLogMessage(AbstractLogger.java:2170)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessageTrackRecursion(AbstractLogger.java:2125)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessageSafely(AbstractLogger.java:2108)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:2002)
	at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:1974)
	at org.apache.logging.log4j.spi.AbstractLogger.info(AbstractLogger.java:1311)
	at org.iota.ict.api.RestApi.onStart(RestApi.java:119)
	at org.iota.ict.utils.RestartableThread$StateTerminated.start(RestartableThread.java:76)
	at org.iota.ict.utils.RestartableThread.start(RestartableThread.java:37)
	at org.iota.ict.utils.RestartableThread$StateTerminated.start(RestartableThread.java:78)
	at org.iota.ict.utils.RestartableThread.start(RestartableThread.java:37)
	at org.iota.ict.Ict.<init>(Ict.java:70)
```

Any idea how to fix this? @jwausle

**Comment by jwausle** [comment]
@JelleMillenaar verify this 

* exist the `./logs` directory - if not create it

or try one of this

1. Try to set cmdline parameter `--log-dir /absolute/path/to/existing/log/dir`
1. Try to set system property `-Dlog4j2.configurationFile=path/to/log4j2.xml` file 

 * You could use [this](https://github.com/iotaledger/ict/blob/master/src/main/resources/log4j2.xml) as base 
 * remove the line 15-32 and line 46.

**Comment by mikrohash** [comment]
creating `./logs` didn't fix it but deleting lines 15-32 did

**Pull request merged**